src/isa/isainfo.h

```diff
@@ -47,28 +44,46 @@ struct RegInfoBase {
   virtual bool regIsReadOnly(unsigned i) const = 0;
 };
 
+/// An index into a single register.
+struct RegIndex {
+  const std::shared_ptr<const RegFileInfoInterface> file;
+  const unsigned index;
+};
+
+using RegInfoVec = std::vector<std::shared_ptr<const RegFileInfoInterface>>;
+using RegInfoMap =
+    std::map<std::string_view, std::shared_ptr<const RegFileInfoInterface>>;
+
 /// The ISAInfoBase class defines an interface for instruction set information.
 class ISAInfoBase {
 public:
   virtual ~ISAInfoBase(){};
   virtual QString name() const = 0;
   virtual ISA isaID() const = 0;
+  virtual const RegInfoMap &regInfoMap() const = 0;
 
-  std::optional<const RegInfoBase *>
-  regInfo(RegisterFileType regFileType = RegisterFileType::GPR) const {
-    if (auto match = m_regInfos.find(regFileType); match != m_regInfos.end()) {
-      return m_regInfos.at(regFileType).get();
-    } else {
-      return {};
+  const std::set<std::string_view> regFileNames() const {
+    std::set<std::string_view> names;
+    for (const auto &regInfo : regInfoMap()) {
+      names.insert(regInfo.second->regFileName());
     }
+    return names;
   }
-  /// Returns the total number of registers in the instruction set.
-  unsigned regCnt() const {
-    unsigned count = 0;
-    for (const auto &pair : m_regInfos) {
-      count += pair.second->regCnt();
+  RegInfoVec regInfos() const {
+    RegInfoVec regVec;
+    for (const auto &regInfo : regInfoMap()) {
+      regVec.emplace_back(regInfo.second);
+    }
+    return regVec;
+  }
+  std::optional<const RegFileInfoInterface *>
+  regInfo(const std::string_view &regFileName) const {
+    if (auto match = regInfoMap().find(regFileName);
+        match != regInfoMap().end()) {
+      return {regInfoMap().at(regFileName).get()};
+    } else {
+      return {};
     }
```

Comment:
Nit: sometimes it's considered better style to not have returns in else statements (https://clang.llvm.org/extra/clang-tidy/checks/readability/else-after-return.html).

Reply:
Interesting, thanks for the tip!